README.md
Multivariate Generalized Propensity Score (mvGPS)
================

<!-- begin badges -->

[![CRAN
version](http://www.r-pkg.org/badges/version/mvGPS)](https://cran.r-project.org/package=mvGPS)
[![CRAN RStudio mirror
downloads](http://cranlogs.r-pkg.org/badges/grand-total/mvGPS)](https://cran.r-project.org/package=mvGPS)
[![cran
checks](https://cranchecks.info/badges/worst/mvGPS)](https://cran.r-project.org/web/checks/check_results_mvGPS.html)
[![Build
Status](https://travis-ci.org/williazo/mvGPS.svg?branch=master)](https://travis-ci.org/williazo/mvGPS)
[![codecov](https://codecov.io/gh/williazo/mvGPS/branch/master/graph/badge.svg)](https://codecov.io/gh/williazo/mvGPS)
<!-- end badges -->

The goal of this package is to expand currently available software to
estimate weights for multivariate continuous exposures. Weights are
formed assuming a multivariate normal distribution for the simultaneous
exposures.

# Installation

You can install `mvGPS` from CRAN for the stable release or GitHub for
the development version using the following code:

``` r
# stable release in CRAN
install.packages("mvGPS")

# development version on GitHub
install.packages("devtools")
devtools::install_github("williazo/mvGPS")
```

## Example

### Data Generating

To illustrate a simple setting where this multivariate generalized
propensity score would be useful, we can construct a directed acyclic
graph (DAG) with a bivariate exposure, **D**=(D<sub>1</sub>,
D<sub>2</sub>), confounded by a set **C**=(C<sub>1</sub>, C<sub>2</sub>,
C<sub>3</sub>). In this case we assume C<sub>1</sub> and C<sub>2</sub>
are associated with D<sub>1</sub>, while C<sub>2</sub> and C<sub>3</sub>
are associated with D<sub>2</sub> as shown below.

<img src="README_files/figure-gfm/dag_draw-1.png" width="50%" style="display: block; margin: auto;" />

To generate this data we first draw *n*=200 samples from **C** assuming
a multivariate normal distribution with mean equal to zero, variance
equal to 1, and constant covariance of 0.1.

Next we define our exposure as a linear function of our confounders.
Explicitly these two equations are defined as

E\[D<sub>1</sub>|**C**\]=0.5C<sub>1</sub>+C<sub>2</sub>,

E\[D<sub>2</sub>|**C**\]=0.3C<sub>2</sub>+0.75C<sub>3</sub>.

With this construction, the exposures have one confounder in common,
C<sub>2</sub>, and one independent confounder. The effect size of the
confounders vary for each exposure. We assume that the conditional
distribution of **D** given **C** is bivariate normal with conditional
correlation equal to 0.2 and conditional variance equal to 2.

To generate the set of confounders and the corresponding bivariate
exposure we can use the function `gen_D()` as shown below.

``` r
require(mvGPS)
sim_dt <- gen_D(method="u", n=200, rho_cond=0.2, s_d1_cond=2, s_d2_cond=2, k=3,
                C_mu=rep(0, 3), C_cov=0.1, C_var=1,
                d1_beta=c(0.5, 1, 0), d2_beta=c(0, 0.3, 0.75), seed=06112020)
D <- sim_dt$D
C <- sim_dt$C
```

By construction our marginal correlation of D is a function of
parameters from the distribution of **C**, coefficients of conditional
mean equations, and conditional covariance parameter. For the above
specification the true marginal correlation of exposure is equal to 0.24
and our observed marginal correlation is equal to 0.26.

Finally, we specify our outcome, Y, as a linear combination of the
confounders and exposure. The mean of the dose-response equation is
shown below,

E\[Y|**D**,
**C**\]=0.75C<sub>1</sub>+1C<sub>2</sub>+0.6C<sub>3</sub>+D<sub>1</sub>+D<sub>2</sub>.

Both exposures have treatment effect sizes equal to one. The standard
deviation of our outcome is set equal 2.

``` r
alpha <- c(0.75, 1, 0.6, 1, 1)
sd_Y <- 2
X <- cbind(C, D)
Y <- X%*%alpha + rnorm(200, sd=sd_Y)
```

### Generating Weights

With the data generated, we can now use our primary function `mvGPS()`
to estimate weights. These weights are constructed such that the
numerator is equal to the marginal density, with the denominator
corresponding to the conditional density, i.e., the multivariate
generalized propensity score.

![weight equation
generic](https://latex.codecogs.com/png.latex?%5Cinline%20w%3D%5Cfrac%7Bf%28%5Cmathbf%7BD%7D%29%7D%7Bf%28%5Cmathbf%7BD%7D%5Cmid%5Cmathbf%7BC%7D%29%7D)

In our case since the bivariate exposure is assumed to be bivariate
normal, we can break both the numerator and denominator into full
conditional densities knowing that each univariate conditional
expression will remain normally distributed.

![weight equation bivariate
normal](https://latex.codecogs.com/png.latex?%5Cinline%20w%3D%5Cfrac%7Bf%28D_2%7CD_1%29f%28D_1%29%7D%7Bf%28D_2%7CD_1%2CC_2%2CC_3%29f%28D_1%7CC_1%2CC_2%29%7D)

Notice in the equation above, we are also able to specify the
confounding set for each exposure separately.

``` r
require(mvGPS)
out_mvGPS <- mvGPS(D=D, C=list(C[, 1:2], C[, 2:3]))
w <- out_mvGPS$w
```

This vector `w` now can be used to test balance of confounders by
comparing weighted vs. unweighted correlations and to estimate the
treatment effects using weighted least squares regression.

### Balance Assessment

For continuous exposure(s) we can asses balance using several metrics
such as euclidean distance, maximum absolute correlation, and average
absolute correlation where correlation refers to the Pearson correlation
between exposure and covariate.

Below we use the function `bal()` to specify a set of potential models
to use for comparison. Possible models that are available include:
mvGPS, Entropy, CBPS, GBM, and PS. For methods other than mvGPS which
can only estimate univariate continuous exposure, each exposure is fit
separately so that weights are generated for both exposures.

``` r
require(knitr)
bal_results <- bal(model_list=c("mvGPS", "entropy", "CBPS", "PS", "GBM"), D, C=list(C[, 1:2], C[, 2:3]))
bal_summary <- bal_results$bal_metrics
#contains overall summary statistics with respect to balance
bal_summary <-data.frame(bal_summary, ESS=c(bal_results$ess, nrow(D)))
#adding in ESS with last value representing the unweighted case
bal_summary <- bal_summary[order(bal_summary$max_cor), ]

kable(bal_summary[, c("euc_dist", "max_cor", "avg_cor", "ESS", "method")],
      digits=4, row.names=FALSE,
      col.names=c("Euc. Distance", "Max. Abs. Corr.",
                  "Avg. Abs. Corr.", "ESS", "Method"))
```

<table>

<thead>

<tr>

<th style="text-align:right;">

Euc. Distance

</th>

<th style="text-align:right;">

Max. Abs. Corr.

</th>

<th style="text-align:right;">

Avg. Abs. Corr.

</th>

<th style="text-align:right;">

ESS

</th>

<th style="text-align:left;">

Method

</th>

</tr>

</thead>

<tbody>

<tr>

<td style="text-align:right;">

0.0930

</td>

<td style="text-align:right;">

0.0884

</td>

<td style="text-align:right;">

0.0331

</td>

<td style="text-align:right;">

163.8253

</td>

<td style="text-align:left;">

mvGPS

</td>

</tr>

<tr>

<td style="text-align:right;">

0.2568

</td>

<td style="text-align:right;">

0.2145

</td>

<td style="text-align:right;">

0.1137

</td>

<td style="text-align:right;">

159.9284

</td>

<td style="text-align:left;">

GBM\_D1

</td>

</tr>

<tr>

<td style="text-align:right;">

0.2592

</td>

<td style="text-align:right;">

0.2288

</td>

<td style="text-align:right;">

0.1085

</td>

<td style="text-align:right;">

179.8179

</td>

<td style="text-align:left;">

PS\_D1

</td>

</tr>

<tr>

<td style="text-align:right;">

0.3095

</td>

<td style="text-align:right;">

0.2321

</td>

<td style="text-align:right;">

0.1092

</td>

<td style="text-align:right;">

178.8683

</td>

<td style="text-align:left;">

entropy\_D2

</td>

</tr>

<tr>

<td style="text-align:right;">

0.2400

</td>

<td style="text-align:right;">

0.2336

</td>

<td style="text-align:right;">

0.0721

</td>

<td style="text-align:right;">

172.2635

</td>

<td style="text-align:left;">

entropy\_D1

</td>

</tr>

<tr>

<td style="text-align:right;">

0.2843

</td>

<td style="text-align:right;">

0.2400

</td>

<td style="text-align:right;">

0.1236

</td>

<td style="text-align:right;">

184.3871

</td>

<td style="text-align:left;">

CBPS\_D1

</td>

</tr>

<tr>

<td style="text-align:right;">

0.3185

</td>

<td style="text-align:right;">

0.2418

</td>

<td style="text-align:right;">

0.1227

</td>

<td style="text-align:right;">

180.3586

</td>

<td style="text-align:left;">

PS\_D2

</td>

</tr>

<tr>

<td style="text-align:right;">

0.3285

</td>

<td style="text-align:right;">

0.2502

</td>

<td style="text-align:right;">

0.1219

</td>

<td style="text-align:right;">

142.8799

</td>

<td style="text-align:left;">

GBM\_D2

</td>

</tr>

<tr>

<td style="text-align:right;">

0.5009

</td>

<td style="text-align:right;">

0.3168

</td>

<td style="text-align:right;">

0.2421

</td>

<td style="text-align:right;">

200.0000

</td>

<td style="text-align:left;">

unweighted

</td>

</tr>

<tr>

<td style="text-align:right;">

1.1022

</td>

<td style="text-align:right;">

0.6701

</td>

<td style="text-align:right;">

0.5012

</td>

<td style="text-align:right;">

50.2385

</td>

<td style="text-align:left;">

CBPS\_D2

</td>

</tr>

</tbody>

</table>

We can see that our method `mvGPS` achieves the best balance across both
exposure dimensions. In this case we can also note that the effective
sample size after weighting 163.8253 is still sufficiently large that we
not worried about loss of power.

### Bias Reduction

Finally, we want to check that these weights are properly reducing the
bias when we estimate the exposure treatment effect.

``` r
dt <- data.frame(Y, D)
mvGPS_mod <- lm(Y ~ D1 + D2, weights=w, data=dt)
mvGPS_hat <- coef(mvGPS_mod)[c("D1", "D2")]

unadj_hat <- coef(lm(Y ~ D1 + D2, data=dt))[c("D1", "D2")]

bias_tbl <- cbind(truth=c(1, 1), unadj=unadj_hat, mvGPS_hat)
kable(bias_tbl, digits=2, row.names=TRUE,
             col.names=c("Truth", "Unadjusted", "mvGPS"))
```

<table>

<thead>

<tr>

<th style="text-align:left;">

</th>

<th style="text-align:right;">

Truth

</th>

<th style="text-align:right;">

Unadjusted

</th>

<th style="text-align:right;">

mvGPS

</th>

</tr>

</thead>

<tbody>

<tr>

<td style="text-align:left;">

D1

</td>

<td style="text-align:right;">

1

</td>

<td style="text-align:right;">

1.28

</td>

<td style="text-align:right;">

1.12

</td>

</tr>

<tr>

<td style="text-align:left;">

D2

</td>

<td style="text-align:right;">

1

</td>

<td style="text-align:right;">

1.18

</td>

<td style="text-align:right;">

1.05

</td>

</tr>

</tbody>

</table>

To compare the total reduction at bias we look at the total absolute
bias where we see mvGPS has total bias equal to 0.18, or an average
percent bias of 8.79% per exposure, compared to unadjusted total bias
equal to 0.45, or an average percent bias of 22.62% per exposure. We
therefore achieve 2.57 times reduction in bias.

### Defining Estimable Region

An important consideration when using propensity scores to estimate
causal effects are the three key identifying assumptions:

1.  weak ignorability, aka, unconfoundedness, aka, selection on
    observables
2.  stable unit treatment value (SUTVA)
3.  positivity

*Weak ignorability* assumes that the exposure is conditionally
independent of the potential outcomes given the appropriate set of
confounders. Checking balance as shown above is one of the key
diagnostics to determining the legitimacy of this assumption in
practice.

*SUTVA* states that the potential outcome of each unit does not depend
on the exposure that other units receive and that there exists only one
version of each exposure. It is generally an untestable assumption, but
is key to ensuring that the potential outcomes are well-defined and that
the observed outcome given the observed exposure corresponds to the true
potential outcome.

The final identifying assumption, *positivity*, is our focus when
defining estimable regions for multivariate exposure. Positivity posits
that all units have the potential to receive a particular level of
exposure given any value of the confounders. The upshot of this is that
we need to take care when defining the domain of our exposure when
estimating the mvGPS.

Typically in the case of univariate continuous exposure, we often ensure
positivity by restricting the domain to the observed range of exposure
or a trimmed version. A logical extension to the multivariate exposure
would be to define our domain as the product of the range of each
exposure. However, when the exposures of interest are correlated this
domain may not be appropriate. Recall that in our simulated data the
marginal correlation of D<sub>1</sub> and D<sub>2</sub> is 0.26.

Instead, we propose to ensure positivity with multivariate exposures by
defining the domain as the multidimensional convex hull of the observed
exposure values. To obtain the convex hull of our exposure we use the
function `hull_sample()`. This will return the vertices of the convex
hull, and in the case of bivariate exposure it will also sample equally
along a grid of the convex hull and return these values which can be
used for calculating the dose-response surface.

Note that we can also create trimmed versions of either the product of
ranges or convex hull as shown below.

``` r
require(sp)
chull_D <- hull_sample(D)
#generate convex hull of exposure
chull_D_trim <- hull_sample(D, trim_hull=TRUE, trim_quantile=0.95)
#generate trimmed convex hull

bbox_grid <- sp::bbox(chull_D$hpts_vs) #bounding box over convex hull
bbox_df <- data.frame(D1=c(bbox_grid[1, 1], bbox_grid[1, 2],
                           bbox_grid[1, 2], bbox_grid[1, 1]),
                      D2=c(bbox_grid[2, 1], bbox_grid[2, 1],
                           bbox_grid[2, 2], bbox_grid[2, 2]))
bbox_grid_trim <- sp::bbox(chull_D_trim$hpts_vs) #bounding box over trimmed convex hull
bbox_df_trim <- data.frame(D1=c(bbox_grid_trim[1, 1], bbox_grid_trim[1, 2],
                                bbox_grid_trim[1, 2], bbox_grid_trim[1, 1]),
                           D2=c(bbox_grid_trim[2, 1], bbox_grid_trim[2, 1],
                                bbox_grid_trim[2, 2], bbox_grid_trim[2, 2]))
chull_plot <- ggplot(data=data.frame(D), aes(x=D1, D2))+
    geom_point()+
    geom_polygon(data=data.frame(chull_D$hpts_vs), color="indianred4", fill=NA)+
    geom_polygon(data=data.frame(chull_D_trim$hpts_vs), color="indianred1", fill=NA, alpha=0.4)+
    geom_polygon(data=bbox_df, color="dodgerblue4", fill=NA)+
    geom_polygon(data=bbox_df_trim, color="dodgerblue1", fill=NA, alpha=0.4)+
    xlab("D1")+
    ylab("D2")+
    theme_bw()
chull_plot
```

<img src="README_files/figure-gfm/chull_plot-1.png" width="50%" style="display: block; margin: auto;" />

In dark red we have the observed convex hull and in light red we have
the trimmed convex hull at the 95th percentile. In dark blue we have the
observed product range and in light blue we have the trimmed product
range at the 95th percentile.

Notice that by trimming we are further restricting our domains to high
density regions of the exposure. We can also see that by restricting to
the convex hull we are avoiding areas with sparse data that are included
in the product range domain.

### Dose-Response Surface

When exposure is bivariate, the resulting dose-response function is a
surface. Using the weighted regression model described above to
incorporate the weights, we can predict across our convex hull domain to
gain intuition about how altering the exposures effects the outcome of
interest. To see an example of this type of dose-response surface on an
application to analyzing obesity intervention programs in Los Angeles
County visit <https://williazo.github.io/resources/>.


NEWS
version 1.2.2 (2021-12-05)
  + Remove hyperlink LaTex equations from README to prevent errors

version 1.2.1 (2021-10-16)
  + Update DESCRIPTION to fix windows build errors

version 1.2.0 (2021-08-11)
  + If trimming, only trim weights at upper quantile

version 1.0.2 (2020-09-17)
  + Package is available in CRAN repository <https://cran.r-project.org/web/checks/check_results_mvGPS.html>

version 1.0.0 (2020-08-31)
  + Manuscript available on arXiv and initial CRAN package submission

version 0.0.* (2020-06-05)
  + This is package is under active development.
  + Check back later for further details.


tests/testthat.R
library(testthat)
library(mvGPS)

test_check("mvGPS")


tests/testthat/test-bal.R
set.seed(1515)
D <- matrix(unlist(lapply(seq_len(3), function(m) rnorm(80))), nrow=80)
C1 <- matrix(unlist(lapply(seq_len(3), function(m) rnorm(80))), nrow=80)
C2 <- matrix(unlist(lapply(seq_len(2), function(m) rnorm(80))), nrow=80)
C3 <- matrix(unlist(lapply(seq_len(1), function(m) rnorm(80))), nrow=80)
C <- list(C1, C2, C3)

dt <- data.frame(D, C1, C2, C3)
names(dt) <- paste0("X", seq_len(ncol(dt)))
test_that("Argument check", {
    #unused model arguments warning
    expect_warning(bal(model_list=c("mvGPS", "kernel", "super"), D, C))
    
    #checking fuzzy matching for arguments
    out <- bal(model_list=c("mv", "ent", "PS"), D=D, C=C)
    expect_equal(out$models, c("mvGPS", "entropy", "PS"))
    expect_named(out, c("W", "cor_list", "bal_metrics", "ess", "models"))
    
    out_trim <- bal(model_list=c("mv", "ent", "PS"), D=D, C=C, trim_w=TRUE)
    
    #checking the max trimming
    expect_equal(unname(unlist(lapply(out$W, quantile, 0.99))),
                 unname(unlist(lapply(out_trim$W, max, 0.99))))

    #for each univariate method we expect them to have metric for each exposure,
    #i.e, 2 * 3 and we have one from mvGPS and one unweighted
    expect_equal(nrow(out$bal_metrics), 1 + 1 + 2 * 3)
    
    #if all_uni=FALSE then we should have only one weight for each method
    out_uni <- bal(model_list=c("mv", "ent", "PS"), D=D, C=C, all_uni=FALSE)
    expect_equal(nrow(out_uni$bal_metrics), 1 + 1 + 2 * 1)
    
    out_uni_trim <- bal(model_list=c("mv", "ent", "PS"), D=D, C=C, 
                        all_uni=FALSE, trim_w=TRUE)
    #checking the max trimming
    expect_equal(unname(unlist(lapply(out_uni$W, quantile, 0.99))),
                 unname(unlist(lapply(out_uni_trim$W, max))))
    
    #there is an additional warning whenever GBM is used
    expect_warning(bal(model_list=c("GBM"), D=D, C=C))
    
    #checking passing options to weightit functions. here we expect the GBM function to not have an error
    expect_warning(bal(model_list=c("GBM"), D=D, C=C, stop.method="p.mean"), regexp=NA)
    
})


tests/testthat/test-genD.R
test_that("Argument check", {
    expect_error(gen_D("o", 200, 0.8, 1, 1, 3, rep(0, 3), 1, rep(0, 3), rep(0, 3)))
    
})
test_that("Performance check", {
    #checking to make sure seed is properly set
    u1<- gen_D(method="u", n=200, rho_cond=0.8, s_d1_cond=1, s_d2_cond=1, k=3, 
               C_mu=rep(0, 3), C_cov=0.2, C_var=1, 
               d1_beta=rep(0, 3), d2_beta=rep(0, 3), seed=1)
    u2 <- gen_D(method="u", n=200, rho_cond=0.8, s_d1_cond=1, s_d2_cond=1, k=3, 
                C_mu=rep(0, 3), C_cov=0.2, C_var=1, 
                d1_beta=rep(0, 3), d2_beta=rep(0, 3), seed=1)
    expect_equal(u1$D, u2$D)
    
    #checking that marginal correlation is properly calculated
    zero_rho <-gen_D(method="u", n=200, rho_cond=0, s_d1_cond=1, s_d2_cond=1, k=3, 
                     C_mu=rep(0, 3), C_cov=0, C_var=1, 
                     d1_beta=rep(0, 3), d2_beta=rep(0, 3), seed=1)
    expect_equal(zero_rho$rho, 0)
    
    high_rho <-gen_D(method="u", n=200, rho_cond=0.8, s_d1_cond=1, s_d2_cond=1, k=3, 
                     C_mu=rep(0, 3), C_cov=0, C_var=1, 
                     d1_beta=rep(0, 3), d2_beta=rep(0, 3), seed=1)
    expect_equal(high_rho$rho, 0.8)
    
    #checking names of output
    expect_named(u1, c("D", "C", "D_Sigma", "rho"))
    
    #checking vector normal method
    v1<- gen_D(method="v", n=200, rho_cond=0.8, s_d1_cond=1, s_d2_cond=1, k=3, 
               C_mu=rep(0, 3), C_cov=0.2, C_var=1, 
               d1_beta=rep(0, 3), d2_beta=rep(0, 3), seed=1)
    v2 <- gen_D(method="v", n=200, rho_cond=0.8, s_d1_cond=1, s_d2_cond=1, k=3, 
                C_mu=rep(0, 3), C_cov=0.2, C_var=1, 
                d1_beta=rep(0, 3), d2_beta=rep(0, 3), seed=1)
    expect_equal(v1$D, v2$D)
    
    #checking matrix normal method
    m1<- gen_D(method="m", n=200, rho_cond=0.8, s_d1_cond=1, s_d2_cond=1, k=3, 
               C_mu=rep(0, 3), C_cov=0.2, C_var=1, 
               d1_beta=rep(0, 3), d2_beta=rep(0, 3), seed=1)
    m2 <- gen_D(method="m", n=200, rho_cond=0.8, s_d1_cond=1, s_d2_cond=1, k=3, 
                C_mu=rep(0, 3), C_cov=0.2, C_var=1, 
                d1_beta=rep(0, 3), d2_beta=rep(0, 3), seed=1)
    expect_equal(m1$D, m2$D)
    
})

R/mvGPS.fit.R
#' Fit Weighted Linear Regression for Dose-Response
#' 
#' Fitting the dose-response model for causal inference model using weights
#' generated from various propensity score models. Can use either the 
#' \code{\link[mvGPS]{mvGPS}} function to generate weights or the 
#' \code{\link[mvGPS]{bal}} function to generate entire list of weights from
#' different propensity models.
#'
#' @param formula  object of class \code{formula} which symbolically describes 
#' the dose-response with the outcome on one side and a function of the exposures on
#' the right side.
#' @param W list of weights to use in the dose-response regression
#' @param data data.frame that contains the outcome and covariates for 
#' dose-response. If none is provided then assumes objects are in user's environment.
#'
#' @details 
#' Will describe details here
#' 
#' @importFrom stats poly as.formula
#'
#' @export
mvGPS.fit <- function(formula, W, data = sys.frame(sys.parent())){
    #checks for proper specification of formula
    if(class(formula) != "formula"){
        stop("formula improperly specified. See details", call. = FALSE)
    } else if (length(formula) != 3){
        stop("formula must be 2-sided", call. = FALSE)
    }
    mf <- model.frame(formula, data)
    #outcome vector
    Y <- mf[, 1]
    #number of obs
    n <- length(Y)
    
    W <- as.list(W)
    W_length <- unlist(lapply(W, length))
    if(!all(W_length==n)) stop("all weights in W must be same length as number of units in exposure and confounders", call.=FALSE)

    #fitting an unadjusted model which accounts does not have any weights or control for confounders
    unadj_mod <- lm(mf)
    
    mod_fit <- lapply(W, function(wts) lm(mf, weights=wts))
    #adding the unadjusted model
    mod_fit[["unweighted"]] <- list(mod=unadj_mod)
    
    mod_fit
}


R/bal.R
#' Construct Covariate Balance Statistics for Models with Multivariate Exposure
#'
#'Assessing balance between exposure(s) and confounders is key when performing causal
#'analysis using propensity scores. We provide a list of several models to generate
#'weights to use in causal inference for multivariate exposures, and test the balancing property of these weights
#'using weighted Pearson correlations. In addition, returns the effective sample
#'size.
#'
#' @inheritParams mvGPS
#' @param model_list character string identifying which methods to use when
#' constructing weights. See details for a list of available models
#' @param all_uni logical indicator. If TRUE then all univariate models specified
#' in model_list will be estimated for each exposure. If FALSE will only estimate weights
#' for the first exposure
#' @param ... additional arguments to pass to \code{\link[WeightIt]{weightit}} function
#' if specifying one of these models in the model_list
#'
#' @import WeightIt
#' @import cobalt
#' @import gbm
#' @import CBPS
#'
#' @details
#' When using propensity score methods for causal inference it is crucial to
#' check the balancing property of the covariates and exposure(s). To do this in
#' the multivariate case we first use a weight generating method from the available
#' list shown below.
#'
#' \subsection{Methods Available}{
#'
#'   \itemize{
#'    \item "mvGPS": Multivariate generalized propensity score using Gaussian densities
#'    \item "entropy": Estimating weights using entropy loss
#'    function without specifying propensity score \insertCite{tbbicke2020entropy}{mvGPS}
#'    \item "CBPS": Covariate balancing propensity score for continuous treatments
#'    which adds balance penalty while solving for propensity score parameters \insertCite{fong2018}{mvGPS}
#'    \item "PS": Generalized propensity score estimated using univariate Gaussian densities
#'    \item "GBM": Gradient boosting to estimate the mean function of the propensity score,
#'    but still maintains Gaussian distributional assumptions \insertCite{zhu_boosting}{mvGPS}
#'    }
#'
#'    Note that only the \code{mvGPS} method is multivariate and all others are strictly univariate.
#'    For univariate methods weights are estimated for each exposure separately
#'    using the \code{\link[WeightIt]{weightit}} function given the
#'    confounders for that exposure in \code{C} when \code{all_uni=TRUE}. To estimate
#'    weights for only the first exposure set \code{all_uni=FALSE}.
#' }
#'
#' It is also important to note that the weights for each method can be trimmed at
#' the desired quantile by setting \code{trim_w=TRUE} and setting \code{trim_quantile}
#' in \[0.5, 1\]. Trimming is done at both the upper and lower bounds. For further details
#' see \code{\link[mvGPS]{mvGPS}} on how trimming is performed.
#'
#' \subsection{Balance Metrics}{
#'
#' In this package we include three key balancing metrics to summarize balance
#' across all of the exposures.
#'  \itemize{
#'    \item Euclidean distance
#'    \item Maximum absolute correlation
#'    \item Average absolute correlation
#'   }
#'   \emph{Euclidean distance} is calculated using the origin point as reference, e.g. for \code{m=2}
#'   exposures the reference point is \[0, 0\]. In this way we are calculating how far
#'   the observed set of correlation points are from perfect balance.
#'
#'   \emph{Maximum absolute correlation} reports the largest single imbalance between
#'   the exposures and the set of confounders. It is often a key diagnostic as
#'   even a single confounder that is sufficiently out of balance can reduce performance.
#'
#'   \emph{Average absolute correlation} is the sum of the exposure-confounder correlations.
#'   This metric summarizes how well, on average, the entire set of exposures is balanced.
#'  }
#'
#'  \subsection{Effective Sample Size}{
#'
#'  Effective sample size, ESS, is defined as
#'  \deqn{ESS=(\Sigma_i w_i)^{2}/\Sigma_i w_i^2,}
#'  where \eqn{w_i} are the estimated weights for a particular method \insertCite{kish_ess}{mvGPS}.
#'  Note that when \eqn{w=1} for all units that the \eqn{ESS} is equal to the sample size \eqn{n}.
#'  \eqn{ESS} decreases when there are extreme weights or high variability in the weights.
#'  }
#'
#' @importFrom stats quantile
#'
#' @examples
#' #simulating data
#' sim_dt <- gen_D(method="u", n=150, rho_cond=0.2, s_d1_cond=2, s_d2_cond=2,
#' k=3, C_mu=rep(0, 3), C_cov=0.1, C_var=1, d1_beta=c(0.5, 1, 0),
#' d2_beta=c(0, 0.3, 0.75), seed=06112020)
#' D <- sim_dt$D
#' C <- sim_dt$C
#'
#' #generating weights using mvGPS and potential univariate alternatives
#' require(WeightIt)
#' bal_sim <- bal(model_list=c("mvGPS", "entropy", "CBPS", "PS", "GBM"), D,
#' C=list(C[, 1:2], C[, 2:3]))
#'
#' #overall summary statistics
#' bal_sim$bal_metrics
#'
#' #effective sample sizes
#' bal_sim$ess
#'
#' #we can also trim weights for all methods
#' bal_sim_trim <- bal(model_list=c("mvGPS", "entropy", "CBPS", "PS", "GBM"), D,
#' C=list(C[, 1:2], C[, 2:3]), trim_w=TRUE, trim_quantile=0.9, p.mean=0.5)
#' #note that in this case we can also pass additional arguments using in
#' #WeighIt package for entropy, CBPS, PS, and GBM such as specifying the p.mean
#'
#' #can check to ensure all the weights have been properly trimmed at upper and
#' #lower bound
#' all.equal(unname(unlist(lapply(bal_sim$W, quantile, 0.99))),
#' unname(unlist(lapply(bal_sim_trim$W, max))))
#' all.equal(unname(unlist(lapply(bal_sim$W, quantile, 1-0.99))),
#' unname(unlist(lapply(bal_sim_trim$W, min))))
#'
#' @return
#'    \itemize{
#'      \item \code{W}: list of weights generated for each model
#'      \item \code{cor_list}: list of weighted Pearson correlation coefficients for all confounders specified
#'      \item \code{bal_metrics}: data.frame with the Euclidean distance, maximum absolute correlation, and average absolute correlation by method
#'      \item \code{ess}: effective sample size for each of the methods used to generate weights
#'      \item \code{models}: vector of models used
#'      }
#'
#' @references
#'     \insertAllCited{}
#'
#' @export
#'
bal <- function(model_list, D, C, common=FALSE, trim_w=FALSE, trim_quantile=0.99, all_uni=TRUE, ...){
    check_result <- D_C_check(D, C, common)
    assign("D", check_result$D)
    assign("C", check_result$C)

    m <- ncol(D)

    m_list <- as.list(
        match.arg(
            model_list,
            c("mvGPS", "entropy", "CBPS", "PS", "GBM"),
            several.ok=TRUE
        )
    )
    if(any(is.na(pmatch(model_list, unlist(m_list))))){
        exclude_m <- model_list[is.na(pmatch(model_list, unlist(m_list)))]
        warning("The following model(s) specified but not used: ", paste(exclude_m, collapse=", "), call.=FALSE)
    }

    W <- lapply(m_list, function(mod){
       if(mod=="mvGPS"){
           w <- mvGPS(D, C, common = FALSE, trim_w, trim_quantile)
           #NOTE: we force common to be FALSE as we already create C from check in line 135
           w_mvGPS <- list(w$wts)
           names(w_mvGPS) <- "mvGPS"
           w <- w_mvGPS
       } else {
           #all other methods use the weightit function
           if(all_uni==TRUE){
               w <- lapply(seq_len(m), function(d){
                   w <- tryCatch(
                       expr= {
                           w <- weightit(D[, d]~C[[d]], method=mod, ...)
                           w <- w$weights
                           if(trim_w==TRUE){
                               #trimming the large weights
                               w <- ifelse(w<quantile(w, trim_quantile), w, quantile(w, trim_quantile))
                           }
                           return(w)
                       },
                       error = function(e) {
                           msg <- paste0(
                               mod,
                               " failed estimating exposure ",
                               d,
                               ". Returning 1 for all weights equivalent to no weighting applied."
                              )
                           warning(msg, call. = FALSE)
                           w <- rep(1, nrow(D))
                           return(w)
                       }
                   )
                   w
               })
               names(w) <- paste0(mod, "_", colnames(D))
           } else {
               w <- weightit(D[, 1]~C[[1]], method=mod)
               w <- w$weights
               if(trim_w==TRUE){
                   #trimming the large weights
                   w <- ifelse(w<quantile(w, trim_quantile), w, quantile(w, trim_quantile))
               }
               w <- list(w)
               names(w) <- paste0(mod, "_", colnames(D)[1])
           }
       }
       return(w)
   })
    W <- unlist(W, recursive=FALSE, use.names=TRUE)

    #calculating correlation using weights
    D_corr_unweight <- lapply(seq_len(m), function(dose){
        col_w_corr(C[[dose]], D[, dose])
    })
    names(D_corr_unweight) <- paste0(colnames(D), "_cor")

    cor_list <- lapply(W, function(w){
        D_corr_w <- lapply(seq_len(m), function(dose){
            col_w_corr(C[[dose]], D[, dose], weights=w)
        })
        names(D_corr_w) <- paste0(colnames(D), "_cor")
        return(D_corr_w)
    })
    cor_list[["unweighted"]] <- D_corr_unweight

    #calculating summary balance metrics by group
    bal_metrics <- lapply(cor_list, function(x){
        cor_vec <- unlist(x)
        euc_dist = sum(sqrt(sum(cor_vec^2))) #since point of reference is [0, 0]
        max_cor = max(abs(cor_vec)) #this is equivalent to Chebyshev Distance
        avg_cor = mean(unlist(abs(cor_vec)))
        return(c(euc_dist, max_cor, avg_cor))
    })
    bal_metrics <- do.call(rbind, bal_metrics)
    colnames(bal_metrics) <- c("euc_dist", "max_cor", "avg_cor")
    bal_metrics <- data.frame(bal_metrics, method=row.names(bal_metrics))

    #calculating effective sample size by group
    ess <- lapply(W, function(w) sum(w)^2/sum(w^2))
    ess <- unlist(ess)

    return(list(W=W, cor_list=cor_list, bal_metrics=bal_metrics, ess=ess,
                models=unlist(m_list)))
}
